General Refactoring Design (#163)
cyberark/conjur-api-python3: design/general_refactorings.md

## CLI Refactor

------

The purpose of this document is to outline the parts of the Python CLI/SDK code that we would like to refactor as well as provide a design for each point. Through this refactor, we hope that the code will be more readable, and easier to contribute to and maintain.

### Improve Naming

> ------

Author: Removed

Improve Naming. Go over names of files, functions, folder, classes and rename to more readable names.

> For example `class Resource` should be `class ConjurResource` so code reader would understand it is a conjur resource. Same about the name of that file.+

Author: Added to reccomendation

### Organize directory structure

------

##### Proposal

> Create new stracture of directories divided by the type of the file and their responsibily. For example all control classes in one directory

Author: Done

##### Current State

> Currently directories resemble the following structure:

Author: Done

| Directory Name | Purpose                           |
| :------------- | :-------------------------------- |
| host           | files related to host command     |
| init           | files related to init command     |
| list           | files related to list command     |
| login          | files related to login command    |
| logout         | files related to logout command   |
| policy         | files related to policy command   |
| user           | files related to user command     |
| variable       | files related to variable command |

> And the following files are without parent directories and sit in the main directory. : api.py, client.py, argparse_wrapper.py, cli.py , client.py, config.py, constants.py, credentials_data.py, credentials_form_file.py, endpoint.py, errors.py, http_wrapper.py, resource.py, ssl_service.py, version.py

Author: Done

##### Recommendation

> The change i want to make is to organize the files by their types and not by the command they do and have directories that can be used for the files without directory:

Author: Fixed by your first comment

> The change i want to make is to organize the files by their types and not by the command they do and have directories that can be used for the files without directory:

Author: I meant the parentless files

| Directory Name | Purpose                                                      |
| :------------- | :----------------------------------------------------------- |

> | utils          | wrappers and utils of diffrent python libsssl_servicearg_parse_wrapperhttp_wrapper |

Author: Fixed that point to 
wrappers and utils of diffrent python libs :  arg_parse_wrapper, http_wrapper

> | api            | files that talk directly with conjur.  conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service) |

Author: Done

> | logics         | all logic classes here                                       |

Author: removed

| controllers    | all controller classes                                       |
| logics         | all logic classes                                            |
| data_objects   | all data objects classes                                     |

### Improve Error Handling

------

##### Proposal

> Raise only our own exceptions and not built in or libs exceptions. Helps to better understand the error handling flow

Author: Documented in code

> Raise only our own exceptions and not built in or libs exceptions. Helps to better understand the error handling flow

Author: Upaded the proposal

and allows us to catch our own objects and not general exceptions.  In addition all exception should be documented.

##### Current state

> Through the codebase, we have very general errors that are not representative of the actual error that took place. The exception is not documented

Author: To the function documentation in the code

> For example in `get_certificate` function we do :

Author: Fixed the current state description

> For example in `get_certificate` function we do :

Author: Its not good when we raise general Exception we need to raise our object so we will understand better what error happend and will be able to  `except` it if we want

```python
def get_certificate(self, hostname, port):
    """
    Method for connecting to Conjur to fetch the certificate chain
    """
    if port is None:
        port = DEFAULT_PORT
    try:
        fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
        logging.debug("Successfully fetched certificate")
    except Exception as error:
        raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
                        f"Reason: {str(error)}") from error

    return fingerprint, readable_certificate
```

We raise Exception if anything happens in certificate retrivial.

##### Recommendation

> Go over error and exception handling and create Conjur-specific exceptions. The code block above should be:

Author: Done

```python
    def get_certificate(self, hostname, port):
        """
        Method for connecting to Conjur to fetch the certificate chain
        :raises CertificateRetrivialFailed when certificate retrivial failes
        """
        if port is None:
            port = DEFAULT_PORT
        try:
            fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
            logging.debug("Successfully fetched certificate")

>         except Exception as error:

Author: Yes it would. Can you list which edge cases can happen in 
`fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)`
and which exception can be thrown ?

>             raise CertificateRetrivialFailed(f"Unable to retrieve certificate from {hostname}:{port}. " \

Author: Done

                            f"Reason: {str(error)}") from error

        return fingerprint, readable_certificate
```

> The document improve_error_handling_to_do_list goes over all error handling block where change of the raised object need to be done.

Author: Moved it to appendix

------

### Strong-type function parameters and return types

##### Proposal

Strong-type all parameters and function return types as per Python3 best practices

##### Current state

> Functions are not strongly typed. For example this function:

Author: Done

`def get_variable(self, variable_data):`

##### Recommendation

> Strongly Type the function so it will be like this.

Author: Done

`def get_variable(self, variable_data : VariableData) -> str:`

------

### Explicit Import Statements

##### Proposal

> Make sure all import statement in the code in explicit and involve full path. Its the correct way to import in python and helps readr to see which model we mean if the are more then one class we same name.

Author: Done And Done

##### Current state

> There are implicit import statement in the code

Author: Refactored

```python
from .client import Client
```

##### Recommendation

> Use explicit imnport statements

Author: Done

          raise RuntimeError("You decided not to trust the certificate")
      ```

* `def get_account_info()`

  * `RuntimeError` should be replaced with `MissingParameterException`

    ```python
      try:
          self.init_logic.fetch_account_from_server(self.conjurrc_data)
      except Exception as error:
          logging.warning(f"Unable to fetch the account from the Conjur server. Reason: {error}")
          conjurrc_data.account = input("Enter the Conjur account name (required): ").strip()

          if conjurrc_data.account is None or conjurrc_data.account == '':
              raise RuntimeError("Error: account is required")
      ```

  * `def __ensure_overwrite_file()`

    * `Exception` should be replaced with `ConfirmationException`

    ```python
      force_overwrite = input(f"File {config_file} exists. " \
                            f"Overwrite? (Default=yes): ").strip()
      if force_overwrite != '' and force_overwrite.lower() != 'yes':
        raise Exception(f"Not overwriting {config_file}")
      ```

* init_logic.py

  * `def get_certificate()`

    * `Exception` should be replaced with `FailedToRetrieveCertificateException`

>     *  The following are error can be thrown from `ssl_service.get_certificate` :

Reviewer: ```suggestion
    *  The following are errors that can be thrown from `ssl_service.get_certificate` :
```

      * `TIMEOUT_ERROR` when connection to server fails
      * `socket.gaierror` when there is no such DNS address

> 

Reviewer: we will raise those errors as well or we will make our own objects?

## CLI Refactor

> The purpose of this document is to outline the parts of the Python CLI/SDK code that we would like to refactor as well as provide a design for each point. Through this refactor, we hope that the code will be more readable, and easier to contribute to and maintain.

Reviewer: for each item

"readable and easier to contribute to and maintain" (without the ,)

Author: Fixed

> 

Reviewer: I think we should make a TOC so the reader will see at a high level what you are going to tackle
https://ecotrust-canada.github.io/markdown-toc/

Author: Genereated TOC

##### Recommendation

Go over all naming in the code and rename when need. If we rename  `class Resource` to `class ConjurResource` developer who is reading the code will understand we are referencing a specific resource, a Conjur resource. The same applies for the names of files in the repository.

### Organize directory structure

------

##### Proposal

Update the new current directory structure so that it is divided by the types of files and their responsibilities. For example, all `Controller` classes will be placed in a single directory.

##### Current State

Currently, the directory structure resembles the following

| Directory Name | Purpose                           |
| :------------- | :-------------------------------- |
| host           | files related to host command     |
| init           | files related to init command     |
| list           | files related to list command     |
| login          | files related to login command    |
| logout         | files related to logout command   |
| policy         | files related to policy command   |
| user           | files related to user command     |
| variable       | files related to variable command |

> The following files are without parent directory : api.py, client.py, argparse_wrapper.py, cli.py , client.py, config.py, constants.py, credentials_data.py, credentials_form_file.py, endpoint.py, errors.py, http_wrapper.py, resource.py, ssl_service.py, version.py

Reviewer: without parent directory:

Author: Fixed

##### Recommendation

> I recommend that we organize the files by their types and not by the commands they represent.

Reviewer: let's be consistent with our tone. Sometimes "we" ("we would like to refactor") is written and sometimes "I" ("I recommend that")

Author: Changed I to we where is possible and change we to passive form where is possible

For example this statement changed to 
The files need to be organized by their types and not by the commands they represent.

| Directory Name | Purpose                                                      |
| :------------- | :----------------------------------------------------------- |

> | wrappers       | wrappers and utils of diffrent python libs :  arg_parse_wrapper, http_wrapper |

Reviewer: different Python libraries*

Author: Changed

> | conjur_api     | files that talk directly with conjur. For Example   `conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service)` |

Reviewer: files that communicate directly with Conjur. For example, `conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service)`

I feel like I've commented on these a few times already 🤔

Author: Sorry if im missed. Fixed it now.

| logic          | all logic classes                                            |
| controller     | all controller classes                                       |
| logic          | all logic classes                                            |
| data_objects   | all data objects classes                                     |

### Improve Error Handling

------

##### Proposal

> This will help to better under the error handling flow. This will allow us to catch our own object and not general exceptions that do not proper helpful information to the end user. All Conjur CLI/SDK-specific exceptions should be documented in the funtions that call them.

Reviewer: replace the first and second sentence with. "Improve our error handling to catch our own Conjur-CLI/SDK specific objects and not general exceptions that do not proper helpful information to the end user."

In this part you are presenting your proposal not your recommendation yet

Reviewer: Where should they be documented?

Reviewer: > in the funtions that call them.

functions*

Author: Fixed The proposal and added the part of documenting the recommendation

##### Current state

> Through the codebase, we have general errors that are not representative of the actual error that took place. The exception is not documented in the fucntion documentation in the code.

Reviewer: "In the codebase, "

"The exceptions are not documented in the function documentation in the code"

Reviewer: what type of documentation are you suggesting?

Author: Adding the line 
:raises [ExceptionType] [Explantion of what being raised]

For example
`:raises CertificateRetrivialFailed when certificate retrivial failes`

> For example, in the following `get_certificate` function, we raise a general exception if anything goes wrong during the certificate retrieval operation.

Reviewer: This example should be before the "The exception is not documented in the fucntion documentation in the code." sentence because if not I am assuming you are providing an example about documenting errors but showing me that our errors are not representative

Author: Lets discuss it

Author: Changed sentence to :For example of not representative exception without the docstring please take a look  in the following `get_certificate` function, where general Exception  is being raised if anything goes wrong during the certificate retrieval operation.

```python
def get_certificate(self, hostname, port):
    """
    Method for connecting to Conjur to fetch the certificate chain
    """
    if port is None:
        port = DEFAULT_PORT
    try:
        fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
        logging.debug("Successfully fetched certificate")
    except Exception as error:
        raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
                        f"Reason: {str(error)}") from error

    return fingerprint, readable_certificate
```

> We raise Exception if anything happens in certificate retrivial.

Reviewer: why is that bad? Please add here why

Author: Its not . I think i copy pasted from appendix . Removing this line

##### Recommendation

Map current error and exception handling and create Conjur-specific exceptions. The code block above should be:

```python
    def get_certificate(self, hostname, port):
        """
        Method for connecting to Conjur to fetch the certificate chain

>         :raises CertificateRetrivialFailed when certificate retrivial failes

Reviewer: retrieval

Reviewer: fails

Author: Fixed

        """
        if port is None:
            port = DEFAULT_PORT
        try:
            fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
            logging.debug("Successfully fetched certificate")
        except Exception as error:
            raise CertificateRetrievalFailedException(f"Unable to retrieve certificate from {hostname}:{port}. " \
                            f"Reason: {str(error)}") from error

        return fingerprint, readable_certificate
```

> Appendinx 1 lists the error handling cases where fixes need to be done.

Reviewer: Appendix

Reviewer: Appendix 1: List of error handling cases

Reviewer: "See Appendix 1 for a list of error handling where fixes are required"

Please provide link to the appendix 1. See [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) for how

Author: Added

------

### Strong-type function parameters and return types

##### Proposal

> Strong-type all parameters and function return types as per Python3 best practices

Reviewer: add . at the end of the sentence

Author: Added

##### Current state

Functions are not strongly typed.

For example, `def get_variable(self, variable_data):`

##### Recommendation

> Strongly Type the function so it will resemble `def get_variable(self, variable_data : VariableData) -> str:`"

Reviewer: ```suggestion
Strong type functions so they will resemble `def get_variable(self, variable_data : VariableData) -> str:`"
```

Author: Done

------

### Explicit Import Statements

##### Proposal

> Make sure all import statement is explicit and contains full path. Explicit paths makes it easier to tell exactly what the imported resource is and where it is. Additionally, absolute imports remain valid even if the location of the imported resources changes.

Reviewer: ```suggestion
Make sure all import statements are explicit and contains full path. Explicit paths make it easier to tell exactly what the imported resource is and which module it is apart of. Additionally, absolute imports remain valid even if the location of the imported resources changes.
```

Author: Done

##### Current state

> The codebase contains implicit import statements

Reviewer: ```suggestion
The codebase contains implicit import statements.
```

Author: Done

```python
from .client import Client
```

##### Recommendation

Use explicit import statements instead.

```python
from conjur.client import Client
```

## APPENDIX

### APPENDIX 1: Error Handling

------

> The following is a mapping of builtin and third-party exceptions that are currently raised in Conjur. In this document, I will provide suggestions for Conjur CLI/SDK-specific exceptions we should use instead.

Reviewer: in the Conjur CLI

Author: Fixed

* api.py

  * `def __init__()` - There are a few places that raise similar exceptions. `RuntimeError` can be replaced with `MissingParameterException`

    * Missing Account Name

      ```python
      self._account = account
      if not self._account:
          raise RuntimeError("Account cannot be empty!")
      ```

    * Missing Url

      * ```python
        self._account = account
        if not self._account:
            raise RuntimeError("Account cannot be empty!")
        ```

  * `def login()` - `RuntimeError` can be replaced with `MissingParameterException`

    * login id or password are missing 

      ```python
      if not login_id or not password:
          # TODO: Use custom error
          raise RuntimeError("Missing parameters in login invocation!")
      ```

  * `def authenticate()`

    * login id or password are missing 

      * ```python
        if not self.login_id or not self.api_key:
            # TODO: Use custom error
            raise RuntimeError("Missing parameters in authentication invocation!")
        ```

  * `def rotate_other_api_key()` 

>     * `Exception` should be replaced with `InvalidResourceException`

Reviewer: this is not related to the others so why is it under `RuntimeError` can be replaced with `MissingParameterException`?

Author: This is about this function
`    def rotate_other_api_key(self, resource: Resource):
        """
        This method is used to rotate a user/host's API key that is not the current user.
        To rotate API key of the current user use rotate_personal_api_key
        """
        if resource.type not in ('user', 'host'):
            raise Exception("Error: Invalid resource type")

        # Attach the resource type (user or host)
        query_params = {
            'role': resource.full_id()
        }
        response = invoke_endpoint(HttpVerb.PUT, ConjurEndpoint.ROTATE_API_KEY,
                                   self._default_params,
                                   api_token=self.api_token,
                                   ssl_verify=self._ssl_verify,
                                   query=query_params).text
        return response`

It raises Exception. What RuntimeError you mean?

      ```python
      if resource.type not in ('user', 'host'):
          raise Exception("Error: Invalid resource type")
      ```

* client.py

  * `def __init__()` 

    ```python
    try:
        credentials = CredentialsFromFile(DEFAULT_NETRC_FILE)
        loaded_netrc = credentials.load(loaded_config['url'])
    except netrc.NetrcParseError as netrc_error:
        raise Exception("Error: netrc is in an invalid format. "
                        f"Reason: {netrc_error}") from netrc_error
    except Exception as exception:
        # pylint: disable=line-too-long
        raise RuntimeError("Unable to authenticate with Conjur. Please log in and try again.") from exception
    ```

>     * `raise Exception` should be replaced with `raise CredentialsParseException`

Reviewer: what about `CredentialsParsingException` ?

Author: Changed

    * `raise RuntimeError` should be replaced with `raise AuthenticationFailedException`

    

* credentials_from_file.py

  * `def load()`

    * `raise Exception` should be replaced with `raise NotLoggedInException`

      ```python
      if netrc_obj.hosts == {}:
          raise Exception("You are already logged out")
      ```

      ```python
      if netrc_auth == "":
          raise Exception("You are already logged out")
      ```

* init_controller.py

  * `def get_server_certificate()` function

    * `RuntimeError` should be replaced with `MissingParameterException`

      ```python
      if self.conjurrc_data.appliance_url == '':
          # pylint: disable=raise-missing-from
          raise RuntimeError("Error: URL is required")
      ```

>     * `RuntimeError` should be replaced with `InvalidURLException`

Reviewer: InvalidURLFormatException ?

Author: Changed

      try:
          self.init_logic.fetch_account_from_server(self.conjurrc_data)
      # pylint: disable=broad-except,logging-fstring-interpolation
      except Exception as error:
          # pylint: disable=line-too-long,logging-fstring-interpolation
          logging.warning(f"Unable to fetch the account from the Conjur server. Reason: {error}")
          # If there was a problem fetching the account from the server, we will request one
          conjurrc_data.account = input("Enter the Conjur account name (required): ").strip()
      
          if conjurrc_data.account is None or conjurrc_data.account == '':
              # pylint: disable=raise-missing-from
              raise RuntimeError("Error: account is required")
      ```

  * `def __ensure_overwrite_file()`

    * `Exception` should be replaced with `ConfirmationException`

      ```python
      force_overwrite = input(f"File {config_file} exists. " \
                              f"Overwrite? (Default=yes): ").strip()
      if force_overwrite != '' and force_overwrite.lower() != 'yes':
          raise Exception(f"Not overwriting {config_file}")
      ```

* init_logic.py

  * `def get_certificate()`

>     * `Exception` should be replaced with `FailedToRetrieveCertificateException` 

Reviewer: I would highly recommend that we evaluate what other errors the openssl exception throws. At least evaluate the effort it would require to make this addition
I believe this task is not only about refactoring the names of the exceptions but _improving_ our exception handling in general

Author: I added it as a bullet so the evaluation will happen during the task. Or you mean it is better be done currently in the design effort?

      ```python
      try:
          fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
          logging.debug("Successfully fetched certificate")
      except Exception as error:
          raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
                          f"Reason: {str(error)}") from error
      ```

* login_controller.py

  * `def get_username()`

    * `RuntimeError` should be replaced with `MissingRequiredParameterException`

      ```python
      self.credential_data.login = input("Enter your login name to log into Conjur: ").strip()
      if self.credential_data.login == '':
          # pylint: disable=raise-missing-from
          raise RuntimeError("Error: Login name is required")
      ```

  * `def get_api_key()`

>     * `RuntimeError` should be replaced with `AuthenticationException`

Reviewer: we have AuthenticationException and AuthenticationFailedException. Let's pick one

Author: Picked AuthenticationFailedException

## CLI Refactor

> The purpose of this document is to outline the parts of the Python CLI/SDK code that we would like to refactor as well as provide a design for each point. Through this refactor, for each item we hope that the code will be more readable and easier to contribute to and maintain.

Reviewer: ```suggestion
The purpose of this document is to outline the parts of the Python CLI/SDK code that we would like to refactor as well as provide a design for each. After this refactor, we hope that the code will be more readable and easier to contribute to and maintain.
```

> ### TOC

Reviewer: Table of Contents

------

> - [CLI Refactor](#cli-refactor)

Reviewer: you dont need this title. this is the title of the document. remove

  * [Organize directory structure](#organize-directory-structure)
      - [Proposal](#proposal)
      - [Current State](#current-state)
      - [Recommendation](#recommendation)
  * [Improve Error Handling](#improve-error-handling)
      - [Proposal](#proposal-1)
      - [Current state](#current-state)
      - [Recommendation](#recommendation-1)
  * [Strong-type function parameters and return types](#strong-type-function-parameters-and-return-types)
      - [Proposal](#proposal-2)
      - [Current state](#current-state-1)
      - [Recommendation](#recommendation-2)
  * [Explicit Import Statements](#explicit-import-statements)
      - [Proposal](#proposal-3)
      - [Current state](#current-state-2)
      - [Recommendation](#recommendation-3)
- [APPENDIX](#appendix)
  * [APPENDIX 1: Error Handling](#appendix-1--error-handling)

> <small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>

Reviewer: remove

> Before refactor concrete thing here is a list of things that need to be done all over the code:

Reviewer: I would start with the purpose of this document that you are creating

Please add titles, subtitles, etc. See our template [here](https://github.com/cyberark/secrets-provider-for-k8s/blob/master/design/templates/solution_design.md)

For example:
## CLI Refactor

The purpose of this document is to outline the parts of the Python CLI/SDK code that we would like to refactor as well as provide a design for each point. Through this refactor, we hope that the code will be more readable, and easier to contribute to and maintain, and follow best SOLID practices.

Author: Done

> 1. Go Over All Namings And Improve them

Reviewer: 1. Improve naming

I would also detail what do you mean by naming. Is it class naming? object naming? all of the above? Add them in (). For example (class naming, object naming)

I would also provide a short sentence of how this will help us... what is the benefit?

> 2. Directory Organize:

Reviewer: Organize directory structure

Author: Done

>    Current Directories

Reviewer: The current structure resembles the following...

>    | Directory Name | Porpuse                           |

Reviewer: Purpose

Author: Done

   | :------------- | :-------------------------------- |
   | host           | files related to host command     |
   | init           | files related to init command     |
   | list           | files related to list command     |
   | login          | files related to login command    |
   | logout         | files related to logout command   |
   | policy         | files related to policy command   |
   | user           | files related to user command     |
   | variable       | files related to variable command |

>    

Reviewer: The following files are without parent directories and sit in the main directory...

Author: Done

   And this files are without a folder : api.py, client.py, argparse_wrapper.py, cli.py , client.py, config.py, constants.py, credentials_data.py, credentials_form_file.py, endpoint.py, errors.py, http_wrapper.py, ssl_service.py, version.py

>    The change i want to make is to organize the files by their types and not by the command they do and have directories that can be used for the files without directory:

Reviewer: Recommendation:

Organize the files by their types and not by their command. For example, create Controller, ... folders that will combine all Controllers together regardless of the command they serve

Reviewer: I would organize the doc according to the following 

_Proposal_
Create...

_Recommendation_
Update..

So that each point will be contained in its own subsection

>    | Directory Name | Porpuse                                                      |

Reviewer: Purpose

Author: Fixed

   | :------------- | :----------------------------------------------------------- |

>    | utils          | Wrappers and utils of diffrent python libsssl_servicearg_parse_wrapperhttp_wrapper |

Reviewer: I don't think services should go under utils.. I think they should have their own directory 'Services'

Author: why ssl_service is not ssl_util or ssl_wrapper?
Maybe it should be ssl_client and go to a clients folder with api.py and client.py and rename that folder from api to clients

Reviewer: I would call the folder `api` and have the ssl_client, conjur-client under that folder

>    | api            | Files that do the directly talk with conjurclient.pyapi.py   |

Reviewer: conjurclient.pyapi.py write a bit weird do you mean conjur client.py and api.py

Author: Fixed to Files that do the directly talk with conjur.  client.py, api.py

>    | api            | Files that do the directly talk with conjurclient.pyapi.py   |

Reviewer: files that talk directly with ..

Author: done

>    | logics         | All logic classes here                                       |

Reviewer: lowercase 'a' to follow the pattern

Author: Done

   | controllers    | all controller classes                                       |
   | logics         | all logic classes                                            |
   | data_objects   | all data objects classes                                     |

   

> 3. Go over thrown exceptions in the code and change them to our own exceptions

Reviewer: Error handling

Again, follow the template for readability

## Error handling

_Proposal_
Map current error and exception handling and create Conjur-specific exceptions

_Current state_
Through the codebase, we have very general errors that are not representative of the actual error that took place... <fill in more info>

_Recommendation_
Map out the current error flow and convert general exceptions to Conjur-specific exceptions. This will <explain why this is a great change>

Reviewer: Please provide examples in code. Maybe add Example section under _Current state_

> 4. Type The Function. Use Python3 syntax on all function to add types to the arguments and return types

Reviewer: Strong-type function parameters and return types

Proposal
Strong-type all parameters and function return types as per Python3 best practices.

Current state

Recommendation

Reviewer: Please provide some examples of code where we should do this

Author: gave example from code

                            f"Overwrite? (Default=yes): ").strip()
      if force_overwrite != '' and force_overwrite.lower() != 'yes':
        raise Exception(f"Not overwriting {config_file}")
    ```

* init_logic.py

  * `def get_certificate()`

    * `Exception` should be replaced with `FailedToRetrieveCertificateException`

    *  The following are errors that can be thrown from `ssl_service.get_certificate` :

>       * `TIMEOUT_ERROR` when connection to server fails. We wiil add `except` to it and will raise `ConnectionToConjurFailedException` with  error message indicating that there was no response from the server.

Reviewer: ```suggestion
      * `TIMEOUT_ERROR` when connection to server fails. We will add an `except` clause and raise `ConnectionToConjurFailedException` with an error message indicating that there was no response from the server.
```

>       * `socket.gaierror` when there is no such DNS address. We will add `except` to it and will raise `ConnectionToConjurFailedException` with error message indicating that there is no such DNS address.

Reviewer: ```suggestion
      * `socket.gaierror` when there is no such DNS address. We will add an `except` clause and raise `ConnectionToConjurFailedException` with an error message indicating that there is no such DNS address.
```
